### ISSUE TITLE
sm:passwd($user, $pass) inaccessible to non-dba users

### ISSUE THREAD

**Issue by ahenket** [eXist-db/exist#2150]
So... after 5 years of update troubles that kept us safely at version 2.2, we are now seriously considering 4.3.1 as our next big migration.

When I revisited https://github.com/eXist-db/exist/issues/35 I found the error changed. When you run sm:passwd as non-dba user on your own account you'll get:

`exerr:ERROR org.exist.security.PermissionDeniedException: Permission denied to open collection: /db/system/security/exist/accounts by testuser`

It basically means I'm unable to offer a user to change their own password. Is this regression? At the very least there could have been a better error message in place.

**Comment by duncdrum** [comment]
FYI @JoernT since you recently worked on user-manager features.  
Looks like a regression against #36

**Comment by line-o** [comment]
@ahenket you could still enable users to change their password.
Have a specific script run with a sticky user or group bit set or with executing a code block with `system:as-user`.

**Comment by ahenket** [comment]
> @ahenket you could still enable users to change their password.
> Have a specific script run with a sticky user or group bit set or with executing a code block with `system:as-user`.

We run open source code, so I would need something where the dba password doesn't live in the code. Sure it's doable, thanks for pointing out the suggestion, but really a proper fix is preferable.

**Comment by line-o** [comment]
@ahenket you set the sticky group or user bit in post-install via `xmldb:chmod`. Since the installer runs with the appropriate permissions no password has to given at that point.

**Comment by ahenket** [comment]
There’s no installer at play. There are manually registered users with a password set by me because I’m unable to offer them a way to register themselves and determine their own password

**Comment by line-o** [comment]
You can set the user or group bit in exide as well. 
<img width="948" alt="screenshot 2018-11-05 at 23 06 15" src="https://user-images.githubusercontent.com/657222/48063457-6428c300-e1c5-11e8-9c8b-c86d7a0c7ef4.png">

PS: A password reset script will always need this or something similar, since the application should not run as an administrator and the user cannot be logged in when the password is unknown.

**Comment by JoernT** [comment]
@ahenket i fully agree that this feature is missing in eXist-db.  I have a similar issue in an app and will work on it. Maybe i can come up with a solution but can't promise it within days.

**Comment by adamretter** [comment]
@ahenket I think some things are being confused. I think @line-o means that you should use the `setUid/setGid` bits, NOT the sticky bit. They allow an XQuery to be executed as a different user, so if a .xq file is owner by a user in the DBA group or by the DBA group, and has the setUid/setGid bit, it will execute as though it has DBA permissions, and then you don't need the `system:as-user` call at all.

**Comment by adamretter** [comment]
@ahenket Regards the permissions problem you originally mentioned, can you please tell me the user/group/mode of both `/db/system/security/exist/accounts` and `/db/system/security/exist/accounts/testuser.xml` in your system?

**Comment by adamretter** [comment]
It looks to me like this change happened in 2012 for eXist-db 2.0-RC2 (and was pre-GitHub), see https://github.com/eXist-db/exist/commit/8493b454#diff-e62f7d6102f35961d2376efa40ed5993L40

**Comment by ahenket** [comment]
```sm:get-permissions(xs:anyURI('/db/system/security/exist/accounts'))```
and
```sm:get-permissions(xs:anyURI('/db/system/security/exist/accounts/testuser.xml'))```

    <sm:permission xmlns:sm="http://exist-db.org/xquery/securitymanager" owner="SYSTEM" group="dba" mode="rwxrwx---">
      <sm:acl entries="0"/>
    </sm:permission>

@adamretter Setting the setUid/setGid bit on the xquery like you describe just might work. Maybe that's what @line-o meant too, in which case I just misunderstood. I'll see where that takes me.

**Comment by adamretter** [comment]
So indeed, currently the way eXist-db is setup by default, only a DBA can access user-credentials. You can use a setUid/setGid XQuery to manage password changes though...

Whether the current way eXist-db is setup is certainly up for discussion.

**Comment by dizzzz** [comment]
@adamretter knowing userids is half of the 'attack vector' ; i'd day try to keep it shielded as we server everything via a easily accessible HTTP(s) interface.

**Comment by adamretter** [comment]
@ahenket Can we close this one now?

**Comment by ahenket** [comment]
The default on Linux systems is that users can always change their own password. I think that is an important feature for eXist too. So if the default settings of eXist behave like the default settings in eXist-db 5.2.x, then this ticket may be closed, unless you have a different ticket tracking that change. Could you point me to the tickets that tracks the change if relevant?

**Comment by adamretter** [comment]
> So if the default settings of eXist behave like the default settings in eXist-db 5.2.x

@ahenket I am not sure I follow what you are saying here

**Comment by line-o** [comment]
@adamretter I understand that it should be possible for user **A** to call 
```
sm:passwd('A', 'xxxxxxxxxxxxxxxx')
```
while logged in as **A** and the function to check if 
a) the user is part of DBA **or** 
b) be logged in as that user. 
If true, then change the password.

**Comment by ahenket** [comment]
Default for eXist-db is dba only. I though we were all in agreement that this should be changed the way @line-o stated.

**Comment by adamretter** [comment]
@line-o @ahenket eXist-db already allows a non-DBA user to change their own password - https://github.com/eXist-db/exist/blob/develop/exist-core/src/main/java/org/exist/xquery/functions/securitymanager/AccountManagementFunction.java#L162

So I think we are good.

**Comment by ahenket** [comment]
`sm:passwd('testuser', 'test2')`

yields:

`exerr:ERROR org.exist.security.PermissionDeniedException: Permission denied to open collection: /db/system/security/exist/accounts by testuser [at line 3, column 1]`

I don't think it was fully implemented?
